src/backend/core/api/viewsets.py

```diff
@@ -184,12 +178,69 @@ class UserViewSet(
     mixins.UpdateModelMixin,
     viewsets.GenericViewSet,
 ):
-    """User ViewSet"""
+    """
+    User ViewSet for all interactions with user infos and teams.
+
+    GET /api/users/&q=query
+        Return a list of all users whose email matches the query, completely or partially
+    """
 
     permission_classes = [permissions.IsSelf]
     queryset = models.User.objects.all().select_related("profile_contact")
     serializer_class = serializers.UserSerializer
 
+    def list(self, request, *args, **kwargs):
+        """Limit listed users by a query with throttle protection."""
+        user = self.request.user
+        queryset = self.filter_queryset(self.get_queryset())
+
+        if not user.is_authenticated:
+            return queryset.none()
+
+        # Exclude contacts that:
+        queryset = queryset.filter(
+            # - inactive users
+            is_active=True,
+        )
+
+        # Search by case-insensitive and accent-insensitive trigram similarity
+        if query := self.request.GET.get("q", ""):
+            query = Func(Value(query), function="unaccent")
+            similarity = TrigramSimilarity(
+                Func("email", function="unaccent"),
+                query,
+            )
+            queryset = (
+                queryset.annotate(similarity=similarity)
+                .filter(
+                    similarity__gte=0.01
+                )  # Value lowered to match every email containing the queries
+                .order_by("-similarity")
+            )
+
+        # Throttle protection
+        key_base = f"throttle-user-list-{user.id!s}"
+        key_minute = f"{key_base:s}-minute"
+        key_hour = f"{key_base:s}-hour"
+
+        try:
+            count_minute = cache.incr(key_minute)
+        except ValueError:
+            cache.set(key_minute, 1, 60)
+            count_minute = 1
+
+        try:
+            count_hour = cache.incr(key_hour)
+        except ValueError:
+            cache.set(key_hour, 1, 3600)
+            count_hour = 1
+
+        if count_minute > 20 or count_hour > 150:
+            raise exceptions.Throttled()
```

Comment:
Could we extract this logic part in a dedicated function? It will be reusable and easier to test.

Reply:
Could we implement that maybe ? Seems more the DRH way.
https://www.django-rest-framework.org/api-guide/throttling/

Reply:
I don't know what Samuel had in mind but I feel like this could be sumed up in settings, as suggested in Anto's link : 

```python
REST_FRAMEWORK = {
    'DEFAULT_THROTTLE_CLASSES': [
        'rest_framework.throttling.UserRateThrottle'
    ],
    'DEFAULT_THROTTLE_RATES': {
        'user': '150/hour',
        'user': '20/minute'
    }
}
```

Reply:
no Anon here because Anon is capped at 0/second.